File: src/game/entities/SpriteComponent.java
package game.entities;

import engine.entities.EntityComponent;
import game.GameGraphics;
import game.Sprite;
import game.gl.Texture;

public abstract class SpriteComponent extends EntityComponent {

    private GameGraphics gfx;
    private Sprite sprite;

    public SpriteComponent(
            GameGraphics gfx,
            Texture tex,
            int width,
            int height) {

        this.gfx = gfx;

        sprite = new Sprite(tex, width, height);
    }

    @Override
    public void onSpawn() {
        super.onSpawn();

        gfx.addSprite(sprite);
    }

    @Override
    public void destroy() {
        gfx.removeSprite(sprite);
    }

    /**
     * Gets the x-position at which this Sprite should be rendered, in world
     * units.
     *
     * @return
     */
    protected float getX() {
        return entity.hitbox.x;
    }

    /**
     * Gets the y-position at which this Sprite should be rendered, in world
     * units.
     *
     * @return
     */
    protected float getY() {
        return entity.hitbox.y;
    }

    @Override
    public void update(int delta) {
        super.update(delta);

        // Update Sprite position based on the Entity position
        sprite.x = getX();
        sprite.y = getY();
    }

}

File: src/game/components/sprite.component.ts
import * as PIXI from 'pixi.js';

import { Component } from '../component';
import { Assets } from '../assets';
import { HitboxComponent } from './hitbox.component';

// Utils
import { getHitboxFrom } from '../utils';

export class SpriteComponent extends Component {

  public static readonly KEY = Symbol();

  private sprite: PIXI.Sprite;
  private hitbox: HitboxComponent;

  constructor(
    filename: string,
    spriteSettings?: any,
    tiled?: boolean,
  ) {
    super(SpriteComponent.KEY);

    const texture = Assets.texture(filename);
    this.sprite = tiled
      ? new PIXI.TilingSprite(texture)
      : new PIXI.Sprite(texture);

    if (spriteSettings) {
      this.sprite = Object.assign(this.sprite, spriteSettings);
    }
  }

  public onSpawn(): void {

    // Register this Sprite with Pixi
    const viewport = this.entity.context.getViewport();
    viewport.addChild(this.sprite);
    viewport.sortChildren();

    // Retrieve the Hitbox from the Entity
    this.hitbox = getHitboxFrom(this.entity);

    this.snapToEntity();
  }

  public destroy(): void {
    this.entity.context.getViewport().removeChild(this.sprite);
  }

  public lateUpdate(delta: number): void {
    this.snapToEntity();
  }

  private snapToEntity(): void {
    // Update the position of the Sprite based on the Entity position
    this.sprite.x = this.hitbox.x;
    this.sprite.y = this.hitbox.y;
    this.sprite.width = this.hitbox.width;
    this.sprite.height = this.hitbox.height;
  }

}
